File: packages/expo-brownfield/android/src/main/java/expo/modules/brownfield/SharedState.kt
package expo.modules.brownfield

import expo.modules.kotlin.sharedobjects.SharedObject

class SharedState(val key: String) : SharedObject() {
  private var value: Any? = null

  fun get(): Any? {
    synchronized(this) {
      return value
    }
  }

  fun set(newValue: Any?) {
    synchronized(this) {
      value = newValue
    }
    emit("change", mapOf("value" to newValue))
    BrownfieldState.notifySubscribers(key, newValue)
    BrownfieldState.maybeNotifyKeyRecreated(key)
  }
}

File: packages/expo-brownfield/ios/SharedState.swift
import ExpoModulesCore

public final class SharedState: SharedObject {
  private let lock = NSLock()
  private let key: String
  private var value: Any?

  public init(_ key: String) {
    self.key = key
    super.init()
  }

  public func get() -> Any? {
    lock.lock()
    defer { lock.unlock() }
    return value
  }

  public func set(_ newValue: Any?) {
    lock.lock()
    value = newValue
    lock.unlock()

    emit(event: "change", arguments: ["value": newValue])
    BrownfieldStateInternal.shared.notifySubscribers(key, newValue)
    BrownfieldStateInternal.shared.maybeNotifyKeyRecreated(key)
  }
}
